### ISSUE TITLE
Type narrowing with recursive types generates override errors

### ISSUE THREAD

**Issue by me-and** [python/mypy#15561]
**Bug Report**

Type narrowing when using recursive type aliases seems to fail with an override error.

**To Reproduce**

[Playground link](https://mypy-play.net/?mypy=latest&python=3.11&gist=927c61430d6c35f959845558ed4b0dfc)

```python
from typing import TypeAlias, Union

RecursiveStrList: TypeAlias = Union[str, list['RecursiveStrList']]

class C:
    def f(self) -> RecursiveStrList:
        return ['abc']
        
class D(C):
    def f(self) -> list[str]:  # Return type "list[str]" of "f" incompatible with return type "RecursiveStrList" in supertype "C"  [override]
        return ['abc']
        
class E(C):
    def f(self) -> list[RecursiveStrList]:  # But this is fine
        return ['abc']
```

**Expected Behavior**

I'm fairly certain that the above should produce no errors. I don't want to rule out the possibility that I'm not understanding some of the inheritance rules correctly, but if that were the case, I'd expect there to be errors from both classes `D` and `E` above.

**Actual Behavior**

```
main.py:10: error: Return type "list[str]" of "f" incompatible with return type "RecursiveStrList" in supertype "C"  [override]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

As above, this was reproduced in the playground using Python 3.11, mypy 1.4.0 and the default options.

**Comment by JelleZijlstra** [comment]
It's because `list` is invariant. https://mypy.readthedocs.io/en/stable/common_issues.html#invariance-vs-covariance
